packages/style-engine/src/index.ts

```diff
@@ -0,0 +1,64 @@
+/**
+ * External dependencies
+ */
+import { groupBy, kebabCase } from 'lodash';
+
+/**
+ * Internal dependencies
+ */
+import type {
+	Style,
+	StyleOptions,
+	GeneratedCSSRule,
+	StyleDefinition,
+} from './types';
+import { styleDefinitions } from './styles';
+
+/**
+ * Generates a stylesheet for a given style object and selector.
+ *
+ * @param  style   Style object.
+ * @param  options Options object with settings to adjust how the styles are generated.
+ *
+ * @return generated stylesheet.
+ */
+export function generate( style: Style, options: StyleOptions ): string {
+	const rules = getCSSRules( style, options );
+	const groupedRules = groupBy( rules, 'selector' );
+	const selectorRules = Object.keys( groupedRules ).reduce(
+		( acc: string[], subSelector: string ) => {
+			acc.push(
+				`${ subSelector } { ${ groupedRules[ subSelector ]
+					.map(
+						( rule: GeneratedCSSRule ) =>
+							`${ kebabCase( rule.key ) }: ${ rule.value };`
+					)
+					.join( ' ' ) } }`
+			);
+			return acc;
+		},
+		[]
+	);
+
+	return selectorRules.join( '\n' );
+}
+
+/**
+ * Returns a JSON representation of the generated CSS rules.
+ *
+ * @param  style   Style object.
+ * @param  options Options object with settings to adjust how the styles are generated.
+ *
+ * @return generated styles.
+ */
+export function getCSSRules(
+	style: Style,
+	options: StyleOptions
+): GeneratedCSSRule[] {
+	let rules: GeneratedCSSRule[] = [];
+	styleDefinitions.forEach( ( definition: StyleDefinition ) => {
+		rules = [ ...rules, ...definition.generate( style, options ) ];
```

Comment:
Minor comment: Why do we need immutability here?

Reply:
I don't think it's so much about immutability, but concatenating the array returned from `definition.generate()`. So the following might be roughly equivalent to what we want to do here?

```
rules = rules.concat( definition.generate( style, options ) );
```

If we have a lot of rules, it's potentially a bit faster using `concat` than the spread syntax, too, so happy to change it over.

Reply:
Why not simply add the new rules to the array?

```js
rules.push( ...definition.generate( style, options ) );
```

Reply:
That works, too! 😄
I've pushed an update in https://github.com/WordPress/gutenberg/pull/37978/commits/91109e1a9c4231014f3a1368c39b5c61bcf9fffa